File: packages/shared/src/provenance.ts
/**
 * Provenance — the typed source-attribution object every data-interface return
 * carries (DL-69). Mirrors runtime/app/schemas/provenance.py one-for-one
 * (snake_case, like the rest of packages/shared).
 *
 * Supersedes the bare `extraction_source: "upload"` string the upload-extract
 * tools used to emit. It is a strict superset of what was persisted before, so
 * threading it through tool returns / the case file / findings is additive —
 * it never removes or renames an existing key.
 */

export type ProvenanceSourceKind = 'user_upload' | 'public_data' | 'computed' | 'vendor';

export interface Provenance {
  /**
   * Which adapter produced the value — e.g. "UserUploadedSBC", "UserUploadedEOB",
   * "PlanLibrary", "OneUpHealthCoverage", "EligibilityVendorBenefits".
   */
  adapter: string;
  source_kind: ProvenanceSourceKind;
  /**
   * ISO date/datetime the data is current as-of, or null. REQUIRED (non-null)
   * for AccumulatorSource — accumulator math is freshness-sensitive (DL-69).
   */
  as_of: string | null;
  /** 0.0–1.0 confidence in the value. */
  confidence: number;
  /** Human-readable assumptions the adapter made (e.g. OCR heuristics). */
  assumptions: string[];
}

File: runtime/app/schemas/provenance.py
"""Provenance — typed source attribution for every data-interface return (DL-69).

Mirrors packages/shared/src/provenance.ts one-for-one.

Supersedes the bare ``extraction_source: "upload"`` string the upload-extract
tools used to emit. ``as_of`` is mandatory for AccumulatorSource (accumulator
math is freshness-sensitive); ``AccumulatorResult`` in app/sources/base.py
enforces that. ``Provenance`` is a strict superset of what is persisted today, so
threading it into tool returns / the case file / findings is purely additive.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

ProvenanceSourceKind = Literal["user_upload", "public_data", "computed", "vendor"]


class Provenance(BaseModel):
    adapter: str = Field(description='Adapter that produced the value, e.g. "UserUploadedSBC"')
    source_kind: ProvenanceSourceKind
    as_of: datetime | None = Field(
        default=None,
        description="ISO datetime the data is current as-of. Mandatory for AccumulatorSource (DL-69).",
    )
    confidence: float = Field(ge=0.0, le=1.0, description="0.0–1.0 confidence in the value")
    assumptions: list[str] = Field(default_factory=list)
